### PR TITLE
Driver for ITest from Bilt

### PR DESCRIPTION
**Pull request by edumur**
Allow set and get of voltage with different ramps.

### CODE REVIEW

**Comment by FarBo** [comment]
Hi @edumur 
Thank you for the contribution. 
I will start the review process in a continuous basis.
I will look into why the PR is not passing the CI.

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (astafan8) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:45]: a curiosity/design question - why is it such that the `v` parameter that has the more user-friendly name (similary to other qcodes drivers with `volt` and `voltage` parameters) does NOT actually set the voltage to that value? i'd expect that the v parameter DOES set the voltage and there's some target_v parameter that allows to tell the instrument about one's intent to set the voltage to some value.

another remark - once a voltage is being set via v_and_trig parameter, shouldn't parameter v be updated to have an invalid cache (see `.cache.invalidate` for more info), or even set to some neautral value?

Comment (edumur) [reply qcodes_contrib_drivers/drivers/Bilt/ITest.py:45]: Hi,

Thanks for the remarks.
The trigger story is relative to how the ITest is built.
1. you set a ramp mode.
2. you set a ramp speed.
3. you set a target voltage.
4. you trigger the ramping.

However I discussed with a colleague and he had input that I am going to integrate and come back with a new version.
I will also change the naming to target_voltage which is maybe more meaningful.

**Comment by edumur** [comment]
Hi all,

I remove the trigger thing so now there is only a simple voltage parameter.
I added a fake parameter called "synchronous" which block or not the communication with the instrument.
I updated the example accordingly.

**Comment by edumur** [comment]
Could someone review this driver?
I would like to propose another one after this one?

Thanks!

**Comment by FarBo** [comment]
Hi edmur

Thanks for the reminder. I will review it this week.

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:457]: You are returning self._synchronous[chan_id], but saying return is None.  Please fix ths

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:422]: What are the types of 'i' and 'c'?  I think 'i' is a float and 'c' is an int. If yes, then the return type should be Tuple[float, int]

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:389]: What do you expect if state is not '1' or '0'? can they be other than these two options?

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:261]: The return shoudln't be str?

**Comment by FarBo** [comment]
I put some comments to solve the CI issues. Let's first correct the PR to pass the CI, then I will review the content.

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (edumur) [reply qcodes_contrib_drivers/drivers/Bilt/ITest.py]: Corrected.

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (edumur) [reply qcodes_contrib_drivers/drivers/Bilt/ITest.py]: Corrected.

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (edumur) [reply qcodes_contrib_drivers/drivers/Bilt/ITest.py:389]: I do not expect `state` to be anything different.
However I added an `else` statement with a error in case of.

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (edumur) [reply qcodes_contrib_drivers/drivers/Bilt/ITest.py:261]: Yes indeed !

**Comment by edumur** [comment]
Done, thanks for your input!

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:43]: Doesn't it need any validator for set values?
It can prevent setting out of limit values in software level.

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:61]: validators here?

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:78]: validators here?

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:72]: maybe changing the name to v_range or vrange makes more sense?

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:140]: **kwargs: Any

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:5]: please add Any here

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:163]: You can add **kwargs to this super.__init__ as well?

**Comment by FarBo** [comment]
Thank you for the fix.
I also made some comments in the code. Also, I have a question:
Can pos and neg states be set independently? I ask this because I don't know about the instrument, but I am thinking that setting positive state should disable negative state and vice versa. What do you think?

**Comment by edumur** [comment]
I added your correction.

Concerning your question, I tested both pos and neg request on the device directly and setting one does not disable the other.

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:45]: You can simply say: vals=vals.Numbers(-50, 50)

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:63]: same here

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [reply qcodes_contrib_drivers/drivers/Bilt/ITest.py:81]: Are you sure no validators needed here?

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:110]: Are you sure no validators needed here?

File: docs/examples/ITest.ipynb
Comment (FarBo) [inline docs/examples/ITest.ipynb:158]: Perhaps it's better to split each command in one cell, because in this way, it returns None for dac.ch04.v(0.5), which is misleading.

**Comment by edumur** [comment]
Done!

**Comment by FarBo** [comment]
@edumur 
Thank you for the commits. Could you mark the comments as resolved?

@astafan8 
Could you have a final look?

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (astafan8) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:72]: is it true that the output_mode can only have a certain strings as values? for example, only "ramp" and "non-ramp". If so, then it will be very useful for users and for correctness to use an Enum validator with those strings:
```suggestion
                           set_parser=str,
                           vals=vals.Enum("ramp", "non-ramp")
```

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (astafan8) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:88]: as per this section in the docs https://qcodes.github.io/Qcodes/examples/writing_drivers/Creating-Instrument-Drivers.html#On/Off-parameters , i highly recommend to implement this parameter as suggested below and adjust `_get_chan_state`/`_set_chan_state` methods to accept "on"/"off" as input (and not perform the conversion to them from a boolean):
```suggestion
        self.add_parameter('enabled',
                           docstring='Whether the channel is enabled or not',
                           get_cmd=partial(self._parent._get_chan_state, chan_num),
                           set_cmd=partial(self._parent._set_chan_state, chan_num),
                           val_mapping=create_on_off_val_mapping(on_val="on", off_val="off")
```

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (astafan8) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:119]: this can be included into the definition of the bilt_name parameter with `initial_value=f'Chan{chan_num:02d}'` argument

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (astafan8) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:116]: please refrain from using functions unless it's really needed, see here https://qcodes.github.io/Qcodes/examples/writing_drivers/Creating-Instrument-Drivers.html#Functions, likely simple bound methods would do the job

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (astafan8) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:71]: so far as i see the implementation of the _set_output_function method, apparently output_mode can be also supplies as a float or an int? if that is so, then this parameter should allow that, right? or should that feature be removed from _set_output_function ? if the parameter needs to be able to not only accept a few selected strings but selected numbers, then those numbers can also be added to the `vals.Enum`. or if the parameter is supposed to accept a range of numbers then the `vals.MultiType` validator can be used: `vals.MultiType(vals.Enum("ramp", "foo"), vals.Numbers(0, 2.5))`

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (astafan8) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:130]: is this really needed? :) having a trigger method that just prints its arguments?)) if it's not needed, it's better to remove in order not to confuse the future readers of this code

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (astafan8) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:174]: is there a reason for these being init-only private attributes as opposed to proper parameters that users can change when using the instrument? just in case, a reminder, parameters can have set_cmd=None and get_cmd=None which makes them "in-memory" parameters, so they don't talk to the instrument, but still maintain a value. and then instead of maintaining the `self._synchronous` dictionary on the instrument level, this information can be conveniently maintained on the channel level

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (astafan8) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:320]: shall this raise a proper exception with useful error message? :) otherwise it's less likely that users will understand what is wrong :) something like, `raise ValueError(f"Got unexpected output function mode {mode}")`

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (astafan8) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:363]: whenever a function accepts a union of numbers and strings, the types should be changed to `Union[float, str]` becasue floats are allowed as well (right?)

**Review by astafan8 — APPROVED** [review]
I left some comments, but only a few of them are critical (for example, having the correct types in methods signatures); it's up to you @edumur and @FarBo to judge what's worth changing and what can wait for the future

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [reply qcodes_contrib_drivers/drivers/Bilt/ITest.py:72]: This is recommended.

**Comment by FarBo** [comment]
Thanks Mikhail, I think most of your comments are recommended and it's up to @edumur how he wants to implement them.

@edumur, if you like to consider these comments, please implement them and I will have a final review and approve it. Then we can merge it. Let me know what you prefer.

**Comment by edumur** [comment]
@FarBo I will work on the nice comments of @astafan8 ;) .

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (edumur) [reply qcodes_contrib_drivers/drivers/Bilt/ITest.py:174]: Hum, could point.
I could change it and make in an 'in-memory' thing at the channel level but then how would I access it at the instrument level, for example here:
```python
    def _set_voltage(self, chan:int,
                           v_set:float) -> None:
        """
        Set cmd for the chXX_v parameter
        
        Args:
            chan: The 1-indexed channel number
            v_set: The target voltage
        """
        chan_id = self.chan_to_id(chan)
        self.write('{}VOLT {:.8f}'.format(chan_id, v_set))
        self.write(chan_id + 'TRIG:INPUT:INIT')
        if self._get_chan_synchronous(chan):
            v = self._get_voltage(chan)
            while abs(v_set - v)>=self._synchronous_threshold:
                sleep(self._synchronous_delay)
                v = self._get_voltage(chan)
```

Thanks.

Comment (astafan8) [reply qcodes_contrib_drivers/drivers/Bilt/ITest.py:174]: since the instrument is aware of all the channels and they can be accessed by their name(?), then `self.<chanX>.synchronous_delay()` will just work.

now that i look at this method, it seems to me that it actually also should go inside the channel because there is nothing in this method that depends on the instrument itself. the write/ask method calls on the channel level forward to those on the instrument level, so that you don't have to worry about

**Comment by edumur** [comment]
I, normally, integrated all the comments 🤞.

**Comment by FarBo** [comment]
@edumur 
Could you resolve the comments that are not open?

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py:113]: If you want the instrument is initilized with this parameter, rather than explicitly putting it in the __init__ as a class argument and the creating only readable parameter later, you can create a settable parameter and get it under __init__ like:
def __init__(...):
  .
  .
  .
  self.synchronous_enable()
And don't forget to change synchronous_enable to self.synchronous_enable() anywhere else you used it.

Then if you want it to be initilized with True, you just simply set it to its initial value:
 
```
self.add_parameter('synchronous_enable',
                           docstring='Is the channel in synchronous mode.',
                           get_cmd=None,
                           get_parser=bool,
                           set_cmd=partial(self._parent._set_chan_synchronous, chan_num),
                           initial_value=True
                           vals=vals.Bool())
```

what you have done seems duplication of creating an argument and excatly the same parameter, which is set in init and get as a parameter.

**Comment by FarBo** [comment]
@edumur 

Thanks for the update. I only have one comment, If it is clarified, we can close this PR.

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (edumur) [reply qcodes_contrib_drivers/drivers/Bilt/ITest.py:113]: I wanted the user to have the possibility to change this parameter at the initialization of the device.
Is there another way of doing it?

Comment (FarBo) [reply qcodes_contrib_drivers/drivers/Bilt/ITest.py:113]: If you want to be forced at initialization, then there should not be any default value in it
Then you have it as an argument and pass it to the parameter under init, for example:

Def init(sync: bool):
  Self.enabled_synchronous(sync)


And you remove the initial value from the parameter

Comment (edumur) [reply qcodes_contrib_drivers/drivers/Bilt/ITest.py:106]: I guess it is done.
Please check if it is done in the way you suggested.

File: qcodes_contrib_drivers/drivers/Bilt/ITest.py
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Bilt/ITest.py]: Perhaps chan_num explanation should stay :)

Comment (edumur) [reply qcodes_contrib_drivers/drivers/Bilt/ITest.py]: oops

**Review by FarBo — APPROVED** [review]

**Pull request merged**
